snapcraft/commands/remote.py

```diff
@@ -243,3 +249,63 @@ def _get_effective_base(self) -> str:
             )
 
         return base
+
+    def _get_build_id(self) -> str:
+        """Get the build id for the project.
+
+        The build id is formatted as `snapcraft-<project-name>-<hash>`.
+        The hash is a hash of all files in the project directory.
+
+        :returns: The build id.
+
+        :raises SnapcraftError: If the snapcraft.yaml does not contain a 'name' keyword.
+        """
+        if self._parsed_args.build_id:
+            emit.debug(
+                f"Using build ID {self._parsed_args.build_id!r} passed as a parameter."
+            )
+            return self._parsed_args.build_id
+
+        with open(self._snapcraft_yaml, encoding="utf-8") as file:
+            data = yaml_utils.safe_load(file)
+
+        project_name = data.get("name")
+
+        if project_name:
+            emit.debug(
+                f"Using project name {project_name!r} from "
+                f"{str(self._snapcraft_yaml)!r}."
+            )
+        else:
+            raise SnapcraftError(
+                f"Could not get project name from {str(self._snapcraft_yaml)!r}."
+            )
+
+        project_hash = self._compute_hash()
+
+        build_id = f"snapcraft-{project_name}-{project_hash}"
+        emit.debug(f"Using computed build ID {build_id!r}.")
+        return build_id
+
+    def _compute_hash(self) -> str:
+        """Compute an md5 hash from the contents of the files in the current directory.
+
+        If a file or its contents within the directory are modified, then the hash
+        will be different.
+
+        :returns: A string containing the md5 hash.
+        """
+        # sort the list of files for reproducibility
+        files = sorted([file for file in Path().glob("**/*") if file.is_file()])
+        hashes: List[str] = []
+
+        for file_path in files:
+            md5_hash = md5()  # noqa: S324 (insecure-hash-function)
```

Comment:
I realize that this is copied-over code, but do you happen to know why the hash is built as a hash-of-hashes like this? Instead of a single hash that is updated with the contents of all files, I mean

Reply:
Like this?

```python
md5_hash = md5()  # noqa: S324 (insecure-hash-function)
for file_path in files:
  with open(file_path, "rb") as file:
    # read files in chunks in case they are large
    for block in iter(partial(file.read, 4096), b""):
      md5_hash.update(block)

return md5_hash.hexdigest()  # noqa: S324 (insecure-hash-function)
```

I don't see any reason why it wasn't done like this in the first place.

Reply:
I believe the reason to use concatenated hashes is to make it more collision-resistant (if you just hash all files together you could change file boundaries without affecting the final digest).

Reply:
I left it as-is and added a note to the docstring about the limitations.